### DISCUSSION TITLE
When is comm.Send blocking?

### DISCUSSION THREAD

**Discussion by dosoe** [mpi4py/mpi4py#278] [Q&A] [answered]
The documentation for `MPI.comm.Send` states: 

> This function may block until the message is received. Whether or not [Send](https://mpi4py.readthedocs.io/en/stable/reference/mpi4py.MPI.Comm.html#mpi4py.MPI.Comm.Send) blocks depends on several factors and is implementation dependent

This is unexpected as in the methods summary, `Send` is marked as 'Blocking send'. 
Where could I find more details on when `Send` blocks and when it doesn't?
I'm referring to the stable version as of 7/11/2022. 
Thanks

**Comment by dalcinl** [comment, marked as answer]
This has nothing to do with mpi4py, but with core MPI semantics. You should really read the MPI standard for the full details.

In practice, MPI implementations usually do not block if the message is short, and they block if the message is large, and of course the threshold is implementation defined, and perhaps you can even tweak it at runtime with implementation specific ways like environment variables, configuration files, command line arguments, etc. 

You have to write your code assuming the worst case, that is, that plain `Send()` is blocking and will not return until the matching `Recv()` have been posted.

If you do not feel comfortable with the uncertainty, then you should use for example `Ssend()` (will always block until a matching `Recv()` is posted), or `Isend()` (non blocking). These are not the only options. Again, all the send variants are explained in the MPI standard.

Regarding the apparent mismatch between the method summary and the note admonition, well, once again, just look at the MPI standard, they define `MPI_Send()` as *blocking*, and at a later point they clarify that MPI implementations may actually make it non blocking but users should always write their codes assuming it is blocking.

**Reply by dosoe** [reply]
Thanks, that is very clear
